internal/internal_task_handlers.go

```diff
@@ -939,39 +939,31 @@ ProcessEvents:
 	// the replay of that event will panic on the decision state machine and the workflow will be marked as completed
 	// with the panic error.
 	var nonDeterministicErr error
+	var nonDeterminismType nonDeterminismDetectionType
 	if !skipReplayCheck && !w.isWorkflowCompleted || isReplayTest {
 		// check if decisions from reply matches to the history events
 		if err := matchReplayWithHistory(w.workflowInfo, replayDecisions, respondEvents); err != nil {
 			nonDeterministicErr = err
+			nonDeterminismType = nonDeterminismDetectionTypeReplayComparison
 		}
-	}
-	if nonDeterministicErr == nil && w.err != nil {
-		if panicErr, ok := w.err.(*workflowPanicError); ok && panicErr.value != nil {
-			if _, isStateMachinePanic := panicErr.value.(stateMachineIllegalStatePanic); isStateMachinePanic {
-				if isReplayTest {
-					// NOTE: we should do this regardless if it's in replay test or not
-					// but since previously we checked the wrong error type, it may break existing customers workflow
-					// the issue is that we change the error type and that we change the error message, the customers
-					// are checking the error string - we plan to wrap all errors to avoid this issue in client v2
-					nonDeterministicErr = panicErr
-				} else {
-					// Since we know there is an error, we do the replay check to give more context in the log
-					replayErr := matchReplayWithHistory(w.workflowInfo, replayDecisions, respondEvents)
-					w.wth.logger.Error("Ignored workflow panic error",
-						zap.String(tagWorkflowType, task.WorkflowType.GetName()),
-						zap.String(tagWorkflowID, task.WorkflowExecution.GetWorkflowId()),
-						zap.String(tagRunID, task.WorkflowExecution.GetRunId()),
-						zap.Error(nonDeterministicErr),
-						zap.NamedError("ReplayError", replayErr),
-					)
-				}
-			}
-		}
+	} else if panicErr, ok := w.getWorkflowPanicIfIllegaleStatePanic(); ok {
```

Comment:
this is now preferring a state-machine panic over history-compared-non-determinism which... I don't _think_ is what we want.

iirc: non-determinism triggered earlier could lead to a panic after several steps, and that earlier-non-determinism info is likely more useful as it can tell us "ran activity X instead of Y", which the panic won't know.

Reply:
ah, no, I think that's wrong.  previously it had the possibility of clobbering it, but now it's less likely because it's in an `else` instead of a separate `if`.

rereading more carefully!